lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp

```diff
@@ -106,19 +106,45 @@ class FirRegLower {
 
   void initialize(OpBuilder &builder, RegLowerInfo reg, ArrayRef<Value> rands);
 
-  void addToAlwaysBlock(hw::HWModuleOp module, sv::EventControl clockEdge,
-                        Value clock, std::function<void(OpBuilder &)> body,
+  void createTree(OpBuilder &builder, sv::RegOp reg, Value term, Value next);
+
+  void addToAlwaysBlock(Block *block, sv::EventControl clockEdge, Value clock,
+                        std::function<void(OpBuilder &)> body,
                         ResetType resetStyle = {},
                         sv::EventControl resetEdge = {}, Value reset = {},
                         std::function<void(OpBuilder &)> resetBody = {});
 
+  void addToIfBlock(OpBuilder &builder, Value cond,
+                    const std::function<void()> &trueSide,
+                    const std::function<void()> &falseSide);
+
   using AlwaysKeyType = std::tuple<Block *, sv::EventControl, Value, ResetType,
                                    sv::EventControl, Value>;
   llvm::SmallDenseMap<AlwaysKeyType, std::pair<sv::AlwaysOp, sv::IfOp>>
       alwaysBlocks;
+
+  using IfKeyType = std::pair<Block *, Value>;
+  llvm::SmallDenseMap<IfKeyType, sv::IfOp> ifCache;
 };
 } // namespace
 
+void FirRegLower::addToIfBlock(OpBuilder &builder, Value cond,
+                               const std::function<void()> &trueSide,
+                               const std::function<void()> &falseSide) {
+  sv::IfOp &op = ifCache[std::make_pair(builder.getBlock(), cond)];
+  // always build both sides of the if, in case we want to use an empty else
+  //  later.  This way we don't have to build a new if and replace it.
+  if (!op) {
+    builder.create<sv::IfOp>(cond.getLoc(), cond, trueSide, falseSide);
```

Comment:
Cache update is missing. 
```suggestion
    op = builder.create<sv::IfOp>(cond.getLoc(), cond, trueSide, falseSide);
```